src/server/rpc/procedures/roboscape/firmware/melody.c

```diff
@@ -0,0 +1,269 @@
+#define A5  880
+#define AS5 932
+#define B5  988
+#define C6  1047
+#define CS6 1109
+#define D6  1175
+#define DS6 1245
+#define E6  1319
+#define F6  1397
+#define FS6 1480
+#define G6  1568
+#define GS6 1661
+#define A6  1760
+#define AS6 1865
+#define B6  1976
+#define C7  2093
+#define CS7 2217
+#define D7  2349
+#define DS7 2489
+#define E7  2637
+#define F7  2794
+#define FS7 2960
+#define G7  3136
+#define GS7 3322
+#define A7  3520
+#define AS7 3729
+#define B7  3951
+#define C8  4186
+#define CS8 4435
+#define D8  4699
+#define DS8 4978
+
+//Mario main theme melody
+int mario_melody[] = {
+    E7, E7, 0, E7,
+    0, C7, E7, 0,
+    G7, 0, 0,  0,
+    G6, 0, 0, 0,
+
+    C7, 0, 0, G6,
+    0, 0, E6, 0,
+    0, A6, 0, B6,
+    0, AS6, A6, 0,
+
+    G6, E7, G7,
+    A7, 0, F7, G7,
+    0, E7, 0, C7,
+    D7, B6, 0, 0,
+
+    C7, 0, 0, G6,
+    0, 0, E6, 0,
+    0, A6, 0, B6,
+    0, AS6, A6, 0,
+
+    G6, E7, G7,
+    A7, 0, F7, G7,
+    0, E7, 0, C7,
+    D7, B6, 0, 0
+};
+//Mario main them tempo
+int mario_tempo[] = {
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+
+    9, 9, 9,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+
+    9, 9, 9,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+    12, 12, 12, 12,
+};
+
+//Underworld melody
+int underworld_melody[] = {
+    C4, C5, A3, A4,
+    AS3, AS4, 0,
+    0,
+    C4, C5, A3, A4,
+    AS3, AS4, 0,
+    0,
+    F3, F4, D3, D4,
+    DS3, DS4, 0,
+    0,
+    F3, F4, D3, D4,
+    DS3, DS4, 0,
+    0, DS4, CS4, D4,
+    CS4, DS4,
+    DS4, GS3,
+    G3, CS4,
+    C4, FS4, F4, E3, AS4, A4,
+    GS4, DS4, B3,
+    AS3, A3, GS3,
+    0, 0, 0
+};
+//Underwolrd tempo
+int underworld_tempo[] = {
+    12, 12, 12, 12,
+    12, 12, 6,
+    3,
+    12, 12, 12, 12,
+    12, 12, 6,
+    3,
+    12, 12, 12, 12,
+    12, 12, 6,
+    3,
+    12, 12, 12, 12,
+    12, 12, 6,
+    6, 18, 18, 18,
+    6, 6,
+    6, 6,
+    6, 6,
+    18, 18, 18, 18, 18, 18,
+    10, 10, 10,
+    10, 10, 10,
+    3, 3, 3
+};
+
+int music1_melody[] = { C, b, g, C, b, e};
+int music1_beats[] = { 16, 16, 16,  8,  8,  16};
+
+int music2_melody[] = { C, b, g, C, b, e, R, C, c, g, a, C, R };
+int music2_beats[] = { 16, 16, 16,  8,  8,  16, 32, 16, 16, 16, 8, 8, 8 };
+
+void play_melody(int melody[], int beats[], int length)
+{
+    long tempo = 10;
+    // Initialize core variables
+    int tone_ = 0;
+    int beat = 0;
+    long duration  = 0;
+
+    for (int i=0; i<length; i++) {
+        tone_ = melody[i];
+        beat = beats[i];
+        duration = beat * tempo; // Set up timing
+        print("buzzing freq %d, duration %d, notenumber %d/%d\n", tone_, duration, i, length);
+        freqout(PIEZO_SPEAKER_PIN, duration, tone_);
+        // A pause between notes...
+        int delayTime = duration * 0.3;
```

Comment:
I would keep all the melody and beets in a single array which is terminated by a zero. So this is a single int array, first item is frequency, second is the beet. If the frequency is zero then the melody stops. That would be way simpler.

Reply:
mmm If I'm not mistaken we have frequency 0 in the array of frequencies for some of the melodies. for example line 185. but it makes sense to have a single array doing both. 

I didn't compose the different melodies and they were separate from the source hence the separate arrays of beats and frequencies.